src/crafting.cpp

```diff
@@ -1565,6 +1587,29 @@ void player::complete_craft()
         skill_dice -= paws_rank_penalty * 4;
     }
 
+	/* we select the components we want to use */
+	std::vector<comp_selection> comp_selections = std::vector<comp_selection>();
+	for (auto &it : making->requirements.components) {
+		comp_selection cs;
+		if ((cs = select_component(it, batch_size, true)).use_from == usage::cancel) {
+			/* we canceled componenet selection, so we cancel crafting */
+			activity.type = ACT_NULL;
+			return;
+		}
+		comp_selections.push_back(cs);
+	}
+
+	std::vector<tool_selection> tool_selections = std::vector<tool_selection>();
+	for (auto &it : making->requirements.tools) {
+		tool_selection ts;
+		if ((ts = select_tool(it, batch_size, DEFAULT_HOTKEYS, true)).use_from == usage::cancel) {
```

Comment:
Assignment and comparison (of a member) in an if-condition is kind of unusual and not readable.

You also don't gain anything from it. `ts` has been initialized anyway. Why not split this into one line with the definition and assignment of `ts` and another line for the if statement?